File: crates/defi-core/src/traits/cdp.rs
use alloy::primitives::U256;
use async_trait::async_trait;

use crate::error::Result;
use crate::types::*;

#[async_trait]
pub trait Cdp: Send + Sync {
    fn name(&self) -> &str;
    async fn build_open(&self, params: OpenCdpParams) -> Result<DeFiTx>;
    async fn build_adjust(&self, params: AdjustCdpParams) -> Result<DeFiTx>;
    async fn build_close(&self, params: CloseCdpParams) -> Result<DeFiTx>;
    async fn get_cdp_info(&self, cdp_id: U256) -> Result<CdpInfo>;
}

File: ts/packages/defi-core/src/traits/cdp.ts
import type { OpenCdpParams, AdjustCdpParams, CloseCdpParams, CdpInfo, DeFiTx } from "../types.js";

export interface ICdp {
  name(): string;
  buildOpen(params: OpenCdpParams): Promise<DeFiTx>;
  buildAdjust(params: AdjustCdpParams): Promise<DeFiTx>;
  buildClose(params: CloseCdpParams): Promise<DeFiTx>;
  getCdpInfo(cdpId: bigint): Promise<CdpInfo>;
}
